Improvements to collaboration UI

danilo04 opened the pull request:
Fixes #1508

### Fix
This PR adds a series of improvements asked in #1508. Those improvements include:

1. Update the Collaborator UI to use a FAB to add new collaborators. This is more consistent with other parts of the UI (e.g. tags)

| Empty List      | Collaborators List |
| ----------- | ----------- |
|  ![image](https://user-images.githubusercontent.com/195721/134980512-ec2485f6-da34-4fc5-a29a-c81e15a98943.png)    |    ![image](https://user-images.githubusercontent.com/195721/135012299-030e5efd-1df0-447d-ae1d-7169d1fbe863.png)   |




2. Validate whether the user is trying to add their own email as a collaborator

![image](https://user-images.githubusercontent.com/195721/134980732-5c4d1eae-e74d-4a0d-8bb7-0af36ecd7a89.png)


### Test

1. Go to a note
2. Tap on the menu on the top right corner (dot ...)
3. Tap on collaborators
4. Tap on Add Collaborator and type anything that is not a valid email. A message invalid collaborator should be shown.
5. Write a valid email address and tap on  `Accept`. Should update the layout to list of collaborators
6. Tap on the cross icon to remove the collaborator. Should update the layout to show the initial message
7. Tap on Add Collaborator and enter an email address from another valid Simplenote account
8. The note should be shared with the other account
9. Tap on Add Collaborator and enter your account's email. It should show the error message: `You cannot add yourself as a collaborator`

### Review

Only one developer and one designer are required to review these changes, but anyone can perform the review.


### Release

These changes do not require release notes.

danilo04 committed:
Add empty collaborators as empty view with the same format from tags

danilo04 committed:
Fix warning in activity_tags layout

danilo04 committed:
Add Shared With message as the header of the recycler view

Since we are handling the empty view of the collaborators list with an
empty view, the Shared With message is now the header of a recycler view
and the data items are transformed to CollaboratorDataItem to submit the
list items.

danilo04 committed:
Move sealed class to bottom with other classes

danilo04 committed:
Add SessionManager to provides the current user (auth or unauthorized)

danilo04 committed:
Validate case in which the user tries to add themselves as collaborator

danilo04 committed:
Change error message

danilo04 committed:
Add blank end line for git

danilo04 committed:
Add . to comments

danilo04 committed:
Add test to validate whether the user is adding his account

danilo04 committed:
Add 4dp on image button to make a total of 16dp from right side

ParaskP7 commented on Simplenote/src/main/java/com/automattic/simplenote/authentication/SessionManager.kt:
```diff
@@ -0,0 +1,21 @@
+    data class AuthorizedUser(val user: User) : UserSession()
+}
+
+class SessionManager @Inject constructor(private val simperium: Simperium) {
```
Question (❓): Is it possible to add a test suite to accompany this `SessionManager` class?

danilo04 replied:
Sure. Added in c0c9db1a3608c9121085311ce1fb3f661c0c13eb.

ParaskP7 replied:
Awesome @danilo04 , many thanks and it lots good! 🚀

ParaskP7 commented on Simplenote/src/main/java/com/automattic/simplenote/authentication/SessionManager.kt:
```diff
@@ -0,0 +1,21 @@
+    fun getCurrentUser(): UserSession {
+        val currentUser = simperium.user ?: return UserSession.UnauthorizedUser
+
+        return when(currentUser.email != null && !currentUser.needsAuthorization()) {
```
1. Minor (🔍): Consider running reformat on this class. In my case the `when` is separated by the parenthesis with a space. PS: Same applies to the `when` within the `AddCollaboratorViewModel.addCollaborator(...)` and `AddCollaboratorViewModel.isCurrentUser(...)` functions. Also, on `AddCollaboratorViewModel` the formatting is a bit off in general, can you try running reformat on the whole class?
2. Minor (🔍): Consider adding an empty line at the end of this class.

PS: Do that for every file created in this PR, including test files.

danilo04 replied:
Thanks. Fixed in 6a898bbae08dcdfd0e979691b9026fa4abba6413.

ParaskP7 commented on Simplenote/src/main/java/com/automattic/simplenote/CollaboratorsActivity.kt:
```diff
@@ -73,8 +76,18 @@ class CollaboratorsActivity : ThemedAppCompatActivity() {
+        buttonAddCollaborator.setOnClickListener { viewModel.clickAddCollaborator() }
+    }
+
+    private fun ActivityCollaboratorsBinding.setEmptyListImage(@DrawableRes image: Int) {
```
Suggestion (💡): I know that this piece of code has been copy-pasted from other parts of the codebase, thus you are just reusing the same pattern here. However, consider splitting this into two functions and remove this hardcoded `-1` value.

```
    private fun ActivityCollaboratorsBinding.showEmptyView(@DrawableRes image: Int) {
        empty.image.visibility = View.VISIBLE
        empty.image.setImageResource(image)
        empty.title.text = getString(R.string.no_collaborators)
        empty.message.visibility = View.VISIBLE
        empty.message.text = getString(R.string.add_email_collaborator_message)
    }

    private fun ActivityCollaboratorsBinding.hideEmptyView() {
        empty.image.visibility = View.GONE
        empty.message.visibility = View.GONE
    }

...

    private fun ActivityCollaboratorsBinding.handleCollaboratorsList(collaborators: List<String>) {
        hideEmptyView()
        val items = listOf(HeaderItem) + collaborators.map { CollaboratorItem(it) }
        (collaboratorsList.adapter as CollaboratorsAdapter).submitList(items)
    }

    private fun ActivityCollaboratorsBinding.handleEmptyCollaborators() {
        showEmptyView(R.drawable.ic_collaborate_24dp)
        (collaboratorsList.adapter as CollaboratorsAdapter).submitList(emptyList())
    }
```

PS: I suggest doing the same for all the other classes.

danilo04 replied:
Yes, this was just lazy on my part. Fixed for `CollaboratorsActivity` in 991134021961f5ff28c287f48df99a656c5cb493. I will address the rest of the code in a later PR.

ParaskP7 replied:
Thanks for doing that and taking care of that in a future PR on all the other places as well. 🚀

ParaskP7 commented on Simplenote/src/main/res/layout/activity_collaborators.xml:
```diff
@@ -1,95 +1,56 @@
-    <include layout="@layout/toolbar" />
+    <include
+        layout="@layout/toolbar">
+    </include>
```
1. Minor (🔍): Try running reformat on this layout file.
2. Minor (🔍): Try collapsing the tags when possible.

PS: Do that for every layout file created in this PR.

danilo04 replied:
Fixed in 9aa1abb289892f2998133c367f2242f060e00600.

ParaskP7 commented on Simplenote/src/main/res/layout/activity_collaborators.xml:
```diff
@@ -1,95 +1,56 @@
+            android:id="@+id/empty"
+            layout="@layout/empty_view_message" />
+
+        <ImageButton
```
Question (❓): I know that this piece of code has been copy-pasted from other parts of the codebase, thus you are just reusing the same pattern here. However, what if the official material `FAB` component was used here instead. Have you considered that or tested what the diff would look like?

PS: I am just asking this question here but I don't expect and actually wouldn't change it in this PR. However, creating an issue to investigation and possibly update it might be a good way to move this forward. Wdyt?

danilo04 replied:
This is something that I have not think of since I just want to keep the screens consistent with the other screens on the app. There are other changes in the layout files that also need some improvements such as naming convention, etc. Thus, let me make a list of things to improve in the views and add this case in it to see if it is feasible for us to change all the `FAB` buttons to the material one.

ParaskP7 replied:
Thanks for even thinking about that @danilo04 ! 🙏

ParaskP7 reviewed (approved):
👋 @danilo04 !

I have reviewed and tested this PR as per the instructions, everything works as expected, good job! 🌟 

I have left a test related question (❓), a suggestion (💡) and a few minor (🔍) comments for you to consider. I am going to approve this PR anyway, since none is blocking. I am **NOT** going to merge this PR yet to give you some time to apply any of my suggestions. However, feel free to ignore them and merge the PR yourself.

PS: I was seeing a lot of issues that could have been fixed prior creating a commit if maybe `Reformat code`, `Rearrange code` and `Optimize imports` were enabled. I am also not sure if your AS settings differ from mine. Can we sync on that so I can understand why that is indeed happening and maybe help you out avoiding those in the future?

danilo04 commented:
Howdy 👋  @ParaskP7. Thanks for the review. Let me address your comments

> PS: I was seeing a lot of issues that could have been fixed prior creating a commit if maybe Reformat code, Rearrange code and Optimize imports were enabled.

Definitely, I will enable these options because without it is it a lot of work to check that in every commit.

SylvesterWilmott commented:
Looks good!

Just a couple of things:

1. In the collaborators list, I think we can align the text in the rows to the left side of the screen instead of using the 72dp inset. This gives more space for the email addresses.
2. For markdown enabled notes, the collaborators menu item should be available while in "preview" mode. Currently, it's only shown in the menu while in "Edit" mode.

danilo04 committed:
Reduce 72dp to 16 dp in collaborator rows

danilo04 committed:
Add collaborators menu to markdown fragment

danilo04 commented:
Hi @SylvesterWilmott, thanks for the review.

> In the collaborators list, I think we can align the text in the rows to the left side of the screen instead of using the 72dp inset. This gives more space for the email addresses.

Fixed in 98189e52ea227c629ef1b7612fd1f97831613944.

<img src="https://user-images.githubusercontent.com/195721/135124686-ac3aab82-a1d2-412c-bdd0-3d4f8a2b15c8.png" width="400" height="800" />

> For markdown enabled notes, the collaborators menu item should be available while in "preview" mode. Currently, it's only shown in the menu while in "Edit" mode.

Fixed in 251e7e136afd1309fa5bfca4142fcd578d99781e.

danilo04 committed:
Reformat new files added in this PR

danilo04 committed:
Refactor show and hiding empty view in recycler view

Avoid hardcoded parameters in method calls and improve code quality

SylvesterWilmott commented:
Looks good! Thanks @danilo04

danilo04 committed:
Reformat layouts files to collapse certain tags

danilo04 committed:
Add tests for SessionManager

danilo04 committed:
Keep old format in case

ParaskP7 reviewed (approved):
👋 @danilo04 !

All LGTM, thanks for applying mine and Sylvester's suggestions, still approved on my part! 🥇
